### ISSUE TITLE
enable_multiple_grants not working as expercted

### ISSUE THREAD

**Issue by siniG** [Snowflake-Labs/terraform-provider-snowflake#1534]
**Provider Version**
v0.56.1

**Terraform Version**

v1.3.0

**Describe the bug**

We are using different terraform scripts to create and grant access to databases, schemas and tables.
As an example, when granting roles access to a table and  enable_multiple_grants is set to true, other roles that were not granted in this specific script are being deleted.

Looking into the code, I don't see any implementation of enable_multiple_grants anywhere in the code. It is mentioned as an input variable, but nothing is being done with it.

**Expected behavior**

We expect that grants created in one terraform script will not override the ones created in another

**Code samples and commands**

**terraform script 1 (main.tf)**

```
resource "snowflake_table" "test" {
  database            = "TEST_DB"
  schema              = "TEST_SCHEMA"
  name                = "CUSTOMERS"
  change_tracking     = false

  column {
    name     = "ID"
    type     = "STRING"
    nullable = false
    comment  = "Unique ID"
  }

  column {
    name     = "NAME"
    type     = "STRING"
    nullable = false
    comment  = "Full Name"
  }

  column {
    name     = "COUNTRY"
    type     = "STRING"
    nullable = false
    comment  = "Country"
  }
}

resource "snowflake_role" "service_a" {
  name    = "SERVICE_A"
}

resource "snowflake_table_grant" "service_a" {
  database_name = snowflake_table.test.database
  schema_name   = snowflake_table.test.schema
  table_name    = snowflake_table.test.name

  privilege = "SELECT"
  roles     = [snowflake_role.service_a.name]

  enable_multiple_grants = true
  with_grant_option = false
}
```

**terraform script 2 (main.tf)**
```
resource "snowflake_table" "read_only" {
  name    = "DATA_READ_ONLY"
}

resource "snowflake_table_grant" "data_read_only" {
  database_name = "TEST_DB"
  schema_name   = "TEST_SCHEMA"
  table_name        = "CUSTOMERS"

  privilege = "SELECT"
  roles     = [snowflake_role.read_only.name]

  enable_multiple_grants = true
  with_grant_option = false
}
```

Message we are receiving is as follows

```
  # snowflake_table_grant.data_read_only["TEST_DB.TEST_SCHEMA.CUSTOMERS"] will be updated in-place
  ~ resource "snowflake_table_grant" "data_read_only" {
        id                     = "TEST_DB|FREE_LISTING|TEST_SCHEMA|CUSTOMERS|DATA_READ_ONLY|true"
      ~ roles                  = [
          + "DATA_READ_ONLY",
          - "SERVICE_A",
        ]
        # (7 unchanged attributes hidden)
    }
```

**Comment by lepagea01** [comment]
Just to add to this ticket.

Indeed, I'm not sure whether the `enable_multiple_grants` property is relevant here. I have a scenario that may seem slightly different initially but I believe the root cause to be the same: I encounter a similar situation that produces the same results regardless of whether this property is set to true or false.

In a nutshell, I have a `snowflake_role_grants` resource that grants (usage of) roles to grantee roles:

```
resource "snowflake_role_grants" "all" {
  for_each = { for grant in var.role_grants : grant.key => grant }

  role_name              = each.value.role
  roles                  = each.value.roles
  enable_multiple_grants = true
}
```

The `role_grants` variable is derived from settings in (JSON/YAML) files. The first time `terraform apply` is executed, everything is fine: role 'X' is granted to roles 'g1' and 'g2', say. If I eventually modify settings and want to grant role 'X' in addition to role 'g3', `terraform apply` will produce expected results but:

1. Every subsequent execution of `terraform plan` will signal an update in-place for resource `snowflake_role_grants.all["X"]`, for adding role 'g3' as grantee of role 'X', similar to what the OP mention.
2. Every subsequent execution of `terraform apply` will execute an unnecessary `grant role "X" to role "g3"` statement in Snowflake.

I did not look into the source code of the provider but could this have something to do with the way ids are generated for grant-type resources?  In my example, the first time `terraform apply` is executed, the id of the `snowflake_role_grants.all["X"]` resource will be `X❄️g1,g2❄️` (cute, these snowflake delimiters replacing pipes, but harder to handle in some scenarios!). Afterwards, when I want to grant role 'X' to role 'g3', the resource id will stay the same, which is fine in itself, but I guess this is where things get confusing for the provider. Of course, if role 'g3' was granted role 'X' from the start, the id of the `snowflake_role_grants.all["X"]` resource would be `X❄️g1,g2,g3❄️` and things would be fine until role 'g4' needs to be granted role 'X'...

In short, updating the `roles` property of a grant-type resource does not seem to run gracefully. And I suspect updating the `users` property of the `snowflake_role_grants` resource has the same behaviour.

**Comment by lepagea01** [comment]
And after reading other tickets, this may actually be the same problem as mentioned in issue #1514.

**Comment by sfc-gh-swinkler** [comment]
Is this still an issue in the latest version?

**Comment by alex-ld** [comment]
`0.64.0` same issue

**Comment by sfc-gh-jcieslak** [comment]
Hey 👋
Closing the issue as it's referring to the deprecated resource and we're not supporting them. Please, try to use the latest provider version with the non-deprecated resources (e.g. snowflake_grant_privileges_to_account_role). If the issue persists on the non-deprecated resource (or the feature is present in Snowflake, but not in the provider), please create a new GitHub issue. The [migration guide](https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/main/docs/technical-documentation/resource_migration.md) may help during migration to the latest resources. Here's a list of the latest, non-deprecated, grant resources:

Resources
[snowflake_grant_privileges_to_database_role](https://registry.terraform.io/providers/Snowflake-Labs/snowflake/latest/docs/resources/grant_privileges_to_database_role)
[snowflake_grant_privileges_to_account_role](https://registry.terraform.io/providers/Snowflake-Labs/snowflake/latest/docs/resources/grant_privileges_to_account_role)
[snowflake_grant_account_role](https://registry.terraform.io/providers/Snowflake-Labs/snowflake/latest/docs/resources/grant_account_role)
[snowflake_grant_database_role](https://registry.terraform.io/providers/Snowflake-Labs/snowflake/latest/docs/resources/grant_database_role)
snowflake_grant_application_role (coming soon)
[snowflake_grant_privileges_to_share](https://registry.terraform.io/providers/Snowflake-Labs/snowflake/latest/docs/resources/grant_privileges_to_share)
snowflake_grant_ownership (coming soon)

Data sources
[snowflake_grants](https://registry.terraform.io/providers/Snowflake-Labs/snowflake/latest/docs/data-sources/grants)

Important to note is that the new resources don't have such a flag as `enable_multiple_grants`. Right now, the new resources have `enable_multiple_grants` "enabled by default". Also, there's no option (yet at least) to turn it off, so `enable_multiple_grants = false` is not available in the new grant resources.
